# Modelling With No Parameter Data

For this notebook I'll be looking at models with the same processing as the previous models, but without the use of the parameter data that was used to generate the samples.

In [1]:
import pickle
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.model_selection import GridSearchCV

import xgboost as xgb
from xgboost.sklearn import XGBClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score

In [2]:
# code to toggle how much of the columns and rows that can be seen

pd.set_option('display.max_columns', None)
# pd.set_option('display.max_rows', None) 
# pd.reset_option('display.max_rows')
# pd.reset_option('display.max_columns')

In [3]:
# importing prepared dataset

data = pd.read_csv('../data/csvs/exercise_df_pre_modelling.csv')

In [4]:
# Dropping the parameter columns

columns_to_drop = ['watch_location', 'crown_orientation', 'ref_xy_rotation', 'rel_baseline_speed', 'max_rel_speed_change', 'wrist_offset_deg', 'avg_rep_time' ]
euler = data.drop(columns = columns_to_drop, axis=1)

In [5]:
euler

exercise  euler_x_1  euler_y_1  euler_z_1  euler_x_2  \
0     ARM_RAISE-DUMBBELL  -0.085816  -1.236302   0.301816   0.042375   
1     ARM_RAISE-DUMBBELL  -0.088119   1.100436  -0.024372   0.003808   
2     ARM_RAISE-DUMBBELL   1.139586  -0.343638   1.662768   1.135147   
3     ARM_RAISE-DUMBBELL   0.107772   1.157058   0.295678   0.110381   
4     ARM_RAISE-DUMBBELL  -1.097028  -0.115172  -1.874880  -0.861851   
...                  ...        ...        ...        ...        ...   
1761                V_UP   2.821015   0.146785  -2.071550   2.785911   
1762                V_UP  -2.865691  -0.239653  -1.880968  -2.851378   
1763                V_UP  -2.949753  -0.249991  -1.600973  -2.917628   
1764                V_UP   2.582625  -0.059933   2.762979   2.520011   
1765                V_UP  -2.739013  -0.246669   0.973192  -2.774814   

      euler_y_2  euler_z_2  euler_x_3  euler_y_3  euler_z_3  euler_x_4  \
0     -1.174680   0.175925   0.168404  -0.899041   0.085828   0.219827   
1      0.749305   0.129690  -0.098057   0.239190   0.230513  -0.185167   
2     -0.334355   1.675068   1.067773  -0.291769   1.714797   0.947730   
3      1.103577   0.311994   0.096461   0.989649   0.354047  -0.012689   
4     -0.135584  -1.983688  -0.642732  -0.100868  -2.082381  -0.248864   
...         ...        ...        ...        ...        ...        ...   
1761   0.132476  -2.134581   2.464893   0.211636  -2.263668   1.516585   
1762  -0.221894  -1.911896  -2.600337  -0.329308  -1.999571  -1.700011   
1763  -0.256503  -1.654894  -2.486600  -0.501285  -1.870018  -1.477908   
1764  -0.031885   2.794892   2.140279   0.043371   2.736161   1.040326   
1765  -0.215316   0.991057  -2.466177  -0.267732   0.974516  -1.326815   

      euler_y_4  euler_z_4  euler_x_5  euler_y_5  euler_z_5  euler_x_6  \
0     -0.437375   0.099239   0.300773  -0.010329   0.129291   0.407900   
1     -0.156804   0.333754  -0.227466  -0.369970   0.382848  -0.262618   
2     -0.243223   1.762350   0.760399  -0.178796   1.819864   0.531367   
3      0.707827   0.432988  -0.178059   0.403103   0.550574  -0.323999   
4      0.004173  -2.233841  -0.017940   0.064673  -2.301083   0.171893   
...         ...        ...        ...        ...        ...        ...   
1761   0.090518  -2.249671   0.800563   0.080138  -2.183109   0.438943   
1762  -0.329776  -2.075900  -0.458881  -0.331948  -2.023105  -0.662448   
1763  -0.526288  -2.198942  -0.399156  -0.315687  -2.438873  -0.441370   
1764  -0.040482   1.945464   0.561546  -0.159316   1.553265   0.595590   
1765  -0.155941   1.247512  -0.631843  -0.285624   1.435956  -0.686903   

      euler_y_6  euler_z_6  euler_x_7  euler_y_7  euler_z_7  euler_x_8  \
0      0.273842   0.182183   0.517961   0.423973   0.253067   0.585252   
1     -0.526357   0.415933  -0.295046  -0.586055   0.428899  -0.307376   
2     -0.094574   1.870510   0.191390   0.055967   1.911666   0.019187   
3      0.149375   0.687801  -0.396937   0.010345   0.771466  -0.460105   
4      0.103863  -2.341409   0.322031   0.121368  -2.360101   0.413165   
...         ...        ...        ...        ...        ...        ...   
1761   0.146128  -2.119975   0.600248   0.095443  -2.099318   0.651816   
1762  -0.247972  -1.963734  -0.718197  -0.377120  -1.922447  -0.862477   
1763  -0.338699  -2.376706  -0.547972  -0.440010  -2.359928  -0.649651   
1764  -0.088152   1.482153   0.657829   0.020717   1.550275   0.734873   
1765  -0.211218   1.551347  -0.845333  -0.254568   1.558807  -0.903287   

      euler_y_8  euler_z_8  euler_x_9  euler_y_9  euler_z_9  euler_x_10  \
0      0.479614   0.292933   0.577434   0.479647   0.276679    0.501138   
1     -0.595273   0.416536  -0.294043  -0.570692   0.374039   -0.264820   
2      0.156185   1.914684  -0.029979   0.181491   1.925526    0.041327   
3     -0.109620   0.845526  -0.449609  -0.082222   0.828605   -0.362523   
4      0.126967  -2.367668   0.401195   0.118987  -2.369424    0.294362   
...         ...        ...        

I will need to encode the exercise column again, however with the removal of all the parameter columns, I don't think any scaling will be necessary as the euler angles are all within a common range, so we will leave that out.

In [6]:
# Encoding categorical columns and saving mapping model 

columns_to_encode = ['exercise']

label_mappings = {}
enc = LabelEncoder()

for column in columns_to_encode:
    enc.fit_transform(euler[column])
    
    euler[f'{column}'] = enc.transform(euler[column])
    label_mappings[column] = dict(zip(enc.classes_, enc.transform(enc.classes_)))

# Display the label mappings
for column, mapping in label_mappings.items():
    print(f"Column: {column}")
    for label, encoded_value in mapping.items():
        print(f"  {label} -> {encoded_value}")
    print()

Column: exercise
  ARM_RAISE-DUMBBELL -> 0
  BICEP_CURL-DUMBBELL -> 1
  BIRD_DOG -> 2
  BURPEE -> 3
  CRUNCHES -> 4
  HAMMER_CURL-DUMBBELL -> 5
  OVERHEAD_PRESS-DUMBBELL -> 6
  TRICEP_KICKBACK-BENT_OVER+SINGLE_ARM-DUMBBELL-LEFT -> 7
  TRICEP_KICKBACK-BENT_OVER+SINGLE_ARM-DUMBBELL-RIGHT -> 8
  V_UP -> 9



In [7]:
# Splitting

X=euler.drop(['exercise'], axis=1)
y=euler['exercise']

In [8]:
# Stratified splitter

stratified_splitter = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)

# Use the splitter to get indices for the train and test sets
for train_index, test_index in stratified_splitter.split(X, y):
    X_train, X_test = X.iloc[train_index], X.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

With having the most luck with the XGBoost model in our first round of modelling, I am going to try that one first this time.

### XGBoost

In [9]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning, module="xgboost")
warnings.filterwarnings("ignore", category=UserWarning, message="`eval_metric` in `fit` method is deprecated*")
warnings.filterwarnings("ignore", category=UserWarning, message="`early_stopping_rounds` in `fit` method is deprecated*")

In [10]:
# parameter set up
params = {
    'objective': 'multi:softmax',
    'num_class': len(y_train.unique()),
    'max_depth': 6,
    'eta': 0.3,
    'subsample':1,
    'colsample_bytree':1
}

n_rounds = 100

In [11]:
# Convert to DMatrix

d_train = xgb.DMatrix(X_train, label=y_train)
d_test = xgb.DMatrix(X_test, label=y_test)

In [12]:
# Setting up the XGBoost model

model_xgb = xgb.train(params, d_train, n_rounds)

In [13]:
y_pred_xgb = model_xgb.predict(d_test)

In [14]:
# Checking accuracy score.

acc = accuracy_score(y_test, y_pred_xgb)
acc

0.9067796610169492

#### XGB with GridSearch

In [15]:
# Define the hyperparameter grid

param_grid = {
    'objective': ['multi:softmax'],
    'max_depth': [3, 4, 5, 6, 7],
    'eta': [0.2, 0.1, 0.01, 0.001],
    'subsample': [0.5, 0.7, 1],
}

# Train the initial XGBoost model
model_xgbc = XGBClassifier()
grid_search_xgb = GridSearchCV(model_xgbc, param_grid, cv=5, scoring='accuracy')

# Fit the grid search to the data
grid_search_xgb.fit(X_train, y_train)

print("Best set of hyperparameters: ", grid_search_xgb.best_params_)

Best set of hyperparameters:  {'eta': 0.1, 'max_depth': 7, 'objective': 'multi:softmax', 'subsample': 0.7}


In [16]:
# Setting up model with tuned parameters

best_params = grid_search_xgb.best_params_
best_xgb_model = xgb.XGBClassifier(**best_params)

best_xgb_model.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eta=0.1, eval_metric=None,
              feature_types=None, gamma=None, grow_policy=None,
              importance_type=None, interaction_constraints=None,
              learning_rate=None, max_bin=None, max_cat_threshold=None,
              max_cat_to_onehot=None, max_delta_step=None, max_depth=7,
              max_leaves=None, min_child_weight=None, missing=nan,
              monotone_constraints=None, multi_strategy=None, n_estimators=None,
              n_jobs=None, num_parallel_tree=None, ...)

In [17]:
# prediction

y_pred = best_xgb_model.predict(X_test)

In [18]:
# obtaining accuracy score

accuracy = accuracy_score(y_test, y_pred)
print("Test Accuracy: ", accuracy)

Test Accuracy:  0.9180790960451978


Almost 92% with some tuning and none of the parameter data is substantially better than I was expecting!

### RandomForest

In [19]:
# setting up basic model

rfc = RandomForestClassifier(random_state = 3)

In [20]:
# fitting the model to the stratified/scaled data

rfc.fit(X_train, y_train)

RandomForestClassifier(random_state=3)

In [21]:
# running the predictions on training and testing data

y_train_pred = rfc.predict(X_train)
y_test_pred = rfc.predict(X_test)

In [22]:
# accuracy results for basic RandomForestClassifier

print('Testing accuracy :' ,accuracy_score(y_test, y_test_pred))

Testing accuracy : 0.8813559322033898


How well both models did with the parameter values being removed is a bit surprising to me. I really expected those numbers to have more of an effect.

#### RFC with GridSearch

In [23]:
# Parameter settings

param_grid = {
    'n_estimators': [10, 50, 100],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['sqrt', 'log2', None]
}

In [24]:
# Using GridSearchCV

grid_search_rfc = GridSearchCV(rfc, param_grid, cv=5, n_jobs=-1)
grid_search_rfc.fit(X_train, y_train)

print("Best Hyperparameters:", grid_search_rfc.best_params_)

best_model = grid_search_rfc.best_estimator_


Best Hyperparameters: {'max_depth': None, 'max_features': 'log2', 'min_samples_leaf': 1, 'min_samples_split': 2, 'n_estimators': 100}


In [25]:
y_train_pred = best_model.predict(X_train)
y_pred_best_model = best_model.predict(X_test)

In [26]:
print('Train Accuracy score :', accuracy_score(y_train, y_train_pred))
print('Test Accuracy score :', accuracy_score(y_test, y_pred_best_model))

Train Accuracy score : 1.0
Test Accuracy score : 0.884180790960452


Tuning helped a tiny bit, but not enough to spend more time on. 88% accuracy with omitting the parameters is surprising to me. 

Next notebook: modelling_by_rep